### PR TITLE
SQS Queue Worker working with interruption signals

### PR DESCRIPTION
**Pull request by milan-zelenka**
:information_source: Introducing QueueWorkerState that will be used for external control of the processing loop

### CODE REVIEW

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 1]
```diff
@@ -156,6 +161,16 @@ public function consumeMessages(callable $consumer, string $queueName, array $pa
         $isLoopIterationsLimitEnabled = $this->consumeLoopIterationsCount !== self::CONSUME_LOOP_ITERATIONS_NO_LIMIT;
 
         while (!$isLoopIterationsLimitEnabled || $loopIterationsCounter < $this->consumeLoopIterationsCount) {
+            if ($this->queueWorkerState->shouldStop()) {
+                // TODO: remove logging after testing
```
Comment (milan-zelenka) [inline src/Queue/AWSSQS/SqsQueueManager.php:165]: TODO :exclamation: :exclamation: :exclamation: 
- [ ] remove todo

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 2]
```diff
@@ -156,6 +161,16 @@ public function consumeMessages(callable $consumer, string $queueName, array $pa
...
+                $this->logger->debug(
+                    'Processing of SQS queue is going to shut down gracefully.',
+                    [
+                        LoggerContextField::JOB_QUEUE_NAME => $prefixedQueueName,
+                    ],
+                );
+                break;
+            }
```
Comment (milan-zelenka) [inline src/Queue/AWSSQS/SqsQueueManager.php:173]: TODO :exclamation: :exclamation: :exclamation: 
- [ ] Remove logging after testing.
- [ ] Move the condition to the loop's condition

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 3]
```diff
@@ -62,6 +65,7 @@ public function __construct(
         SqsClientFactoryInterface $sqsClientFactory,
         S3ClientFactoryInterface $s3ClientFactory,
         MessageKeyGeneratorInterface $messageKeyGenerator,
+        QueueWorkerState $queueWorkerState,
```
Comment (dominikvoda) [inline src/Queue/AWSSQS/SqsQueueManager.php:68]: I don't like the idea having this in constructor... calling QueueManager::stop seems much transparent to me than connect this logic over 3rd party object

**Review by dominikvoda — COMMENTED** [review]
comments...

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 4]
```diff
@@ -156,6 +158,17 @@ public function consumeMessages(callable $consumer, string $queueName, array $pa
...
+                // TODO: remove logging after testing
+                $this->logger->debug(
+                    'Processing of SQS queue is going to shut down gracefully.',
+                    [
+                        LoggerContextField::JOB_QUEUE_NAME => $prefixedQueueName,
+                    ],
+                );
+                break;
+            }
+
```
Comment (ptrzdnk) [inline src/Queue/AWSSQS/SqsQueueManager.php:171]: remove when tested

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 5]
```diff
@@ -309,4 +322,13 @@ private function getPrefixedQueueName(string $queueName): string
...
         return $prefixedQueueName;
     }
+
+
+    public function terminateGracefully(): void
+    {
+        // todo - remove when tested
+        $this->logger->debug('SqsQueueManager::terminateGracefully() called');
+
+        $this->isBeingTerminated = true;
```
Comment (ptrzdnk) [inline src/Queue/AWSSQS/SqsQueueManager.php:332]: todo - remove when tested
```suggestion
        $this->isBeingTerminated = true;
```

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 6]
```diff
@@ -156,6 +158,17 @@ public function consumeMessages(callable $consumer, string $queueName, array $pa
...
+                $this->logger->debug(
+                    'Processing of SQS queue is going to shut down gracefully.',
+                    [
+                        LoggerContextField::JOB_QUEUE_NAME => $prefixedQueueName,
+                    ],
+                );
+                break;
+            }
+
             try {
```
Comment (ptrzdnk) [inline src/Queue/AWSSQS/SqsQueueManager.php:172]:
```suggestion
        while (!$this->isBeingTerminated && (!$isLoopIterationsLimitEnabled || $loopIterationsCounter < $this->consumeLoopIterationsCount)) {

            try {
```

File: src/Queue/AWSSQS/SqsQueueManager.php [hunk 7]
```diff
@@ -346,4 +348,20 @@ private function getJobJson(JobInterface $job): string
...
     }
+
+
+    public function terminateGracefully(): void
+    {
+        $this->writeDebugLog('SqsQueueManager::terminateGracefully() reached');
+
+        $this->isBeingTerminated = true;
+
+        $this->writeDebugLog('SqsQueueManager::isBeingTerminated set to ' . Json::encode($this->isBeingTerminated));
```
Comment (michal-tichy-84) [inline src/Queue/AWSSQS/SqsQueueManager.php:359]: [O] Isn't this redundant?

Comment (ptrzdnk) [reply src/Queue/AWSSQS/SqsQueueManager.php:359]: maybe, but I would rather keep this

**Review by vharnach — APPROVED** [review]

**Pull request merged**